wrappers/pkcs11/pkcs11.go

```diff
@@ -0,0 +1,542 @@
+func (k *Wrapper) Encrypt(_ context.Context, plaintext []byte, opt ...wrapping.Option) (*wrapping.BlobInfo, error) {
+	if plaintext == nil {
+		return nil, fmt.Errorf("given plaintext for encryption is nil")
+	}
+
+	env, err := wrapping.EnvelopeEncrypt(plaintext, opt...)
+	if err != nil {
+		return nil, fmt.Errorf("error wrapping data: %w", err)
+	}
+
+	WrappedKey, err := k.client.EncryptDEK(context.Background(), env.Key)
+	if err != nil {
+		return nil, fmt.Errorf("error encrypting data: %w", err)
+	}
+
+	// Store the current key id.
+	k.currentKeyId.Store(k.keyId)
+
+	ret := &wrapping.BlobInfo{
+		Ciphertext: env.Ciphertext,
+		Iv:         env.Iv,
+		KeyInfo: &wrapping.KeyInfo{
+			KeyId:      k.keyId,
+			WrappedKey: WrappedKey,
+		},
+	}
+
+	return ret, nil
+}
+
+// Decrypt is used to decrypt the ciphertext. This should be called after Init.
+func (k *Wrapper) Decrypt(_ context.Context, in *wrapping.BlobInfo, opt ...wrapping.Option) ([]byte, error) {
+	if in == nil {
+		return nil, fmt.Errorf("given input for decryption is nil")
+	}
+
+	keyBytes, err := k.client.DecryptDEK(context.Background(), in.KeyInfo.WrappedKey)
+	if err != nil {
+		return nil, fmt.Errorf("error decrypting data encryption key: %w", err)
+	}
+
+	envInfo := &wrapping.EnvelopeInfo{
+		Key:        keyBytes,
+		Iv:         in.Iv,
+		Ciphertext: in.Ciphertext,
+	}
+	plaintext, err := wrapping.EnvelopeDecrypt(envInfo, opt...)
+	if err != nil {
+		return nil, fmt.Errorf("error decrypting data: %w", err)
+	}
+
+	return plaintext, nil
+}
+
+func GetKeyTypeFromMech(mech uint) (uint, error) {
+	switch mech {
+	case pkcs11.CKM_RSA_PKCS:
+		return pkcs11.CKK_RSA, nil
+	case pkcs11.CKM_AES_CBC_PAD:
+		return pkcs11.CKK_AES, nil
+	default:
+		return 0, fmt.Errorf("unsupported mechanism: %d", mech)
+	}
+}
+
+func MechanisString(mech uint) string {
+	switch mech {
+	case pkcs11.CKM_RSA_PKCS:
+		return "CKM_RSA_PKCS"
+	case pkcs11.CKM_AES_CBC_PAD:
+		return "CKM_AES_CBC_PAD"
+	default:
+		return "Unknown"
+	}
+}
+
+func IsIvNeeded(mech uint) bool {
+	switch mech {
+	case pkcs11.CKM_AES_CBC_PAD:
+		return true
+	default:
+		return false
+	}
+}
+
+func MechanisFromString(mech string) (uint, error) {
+	switch mech {
+	case "CKM_RSA_PKCS":
+		return pkcs11.CKM_RSA_PKCS, nil
+	case "CKM_AES_CBC_PAD":
+		return pkcs11.CKM_AES_CBC_PAD, nil
+	default:
+		return 0, fmt.Errorf("unsupported mechanism: %s", mech)
+	}
+}
+
+type pkcs11KMS struct {
+	// standard PKCS11 configuration options
+	slot      uint
+	pin       string
+	module    string
+	keyId     string
+	label     string
+	mechanism uint
+}
+
+// EncryptDEK uses the PKCS11 encrypt operation to encrypt the DEK.
+func (kms *pkcs11KMS) EncryptDEK(ctx context.Context, plainDEK []byte) ([]byte, error) {
+	p := pkcs11.New(kms.module)
+	err := p.Initialize()
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize PKCS11: %w", err)
+	}
+
+	defer p.Destroy()
+	defer p.Finalize()
```

Comment:
`p.Destroy()` and `p.Finalize()` should probably be called from `Finalize(...)` on the Wrapper, not per-call, otherwise there will be lots and lots of overhead. 

I'm not sure that this is thread safe to do this way anyways (loading the module per-call); Wrappers try to be concurrency friendly.